engine/src/main/java/io/zeebe/engine/state/appliers/ProcessInstanceElementActivatingApplier.java

```diff
@@ -134,19 +130,14 @@ private void cleanupSequenceFlowsTaken(final ProcessInstanceRecord value) {
         processState.getFlowElement(
             value.getProcessDefinitionKey(), value.getElementIdBuffer(), ExecutableFlowNode.class);
 
-    final var tokensBySequenceFlow =
-        sequenceFlowAnalyzer.determineTakenIncomingFlows(value.getFlowScopeKey(), parallelGateway);
-
-    // cleanup deferred records for sequence flow taken to this parallel gateway
-    tokensBySequenceFlow.forEach(
-        (sequenceFlow, tokens) -> {
-          // before a parallel gateway is activated, each incoming sequence flow of the gateway must
-          // be taken (at least once). if a sequence flow is taken more than once then the redundant
-          // token remains for the next activation of the gateway (Tetris principle)
-          final var firstToken = tokens.get(0);
-          elementInstanceState.removeStoredRecord(
-              value.getFlowScopeKey(), firstToken.getKey(), Purpose.DEFERRED);
-        });
+    if (parallelGateway.getIncoming().size() > 1) {
+      // before a parallel gateway is activated, all incoming sequence flows of the gateway must
+      // be taken at least once. decrement the number of the taken sequence flows for each incoming
+      // sequence flow but keep the remaining numbers for the next activation of the gateway.
+      // (Tetris principle)
```

Comment:
not sure I understand this metaphor "tetris principle"

Reply:
I have the term from Daniel and I like it. 

![image](https://user-images.githubusercontent.com/4305769/116239972-95195b80-a763-11eb-90bc-0de3a8cc5413.png)

A parallel gateway has multiple incoming sequence flows. Image the sequence flow as a token in the flow. When the token reaches the gateway then it checks if all tokens arrived (i.e. if the row is full). Then, it activates the gateway and removes the tokens (clear the row). If a token arrives on the same sequence flow then it waits for the next activation (i.e. build a new level above).